File: LeetCode/Medium/0322-coin-change/0322-coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function (coins, amount) {
    if (amount === 0) return 0;

    const INF = amount + 1;
    const dp = new Array(amount + 1).fill(INF);
    dp[0] = 0;

    for (const c of coins) {
        for (let x = c; x <= amount; x++) {
            // 如果 dp[x - c] 仍是 INF，说明该状态不可达，此时 INF + 1 仍然安全
            dp[x] = Math.min(dp[x], dp[x - c] + 1);
        }
    }
    return dp[amount] === INF ? -1 : dp[amount];
};

File: LeetCode/Medium/0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        
        const int INF = amount + 1;
        vector<int> dp(amount + 1, INF);
        dp[0] = 0;

        for (int i = 0; i < coins.size(); i++) {       // 遍历物品
            for (int j = coins[i]; j <= amount; j++) { // 遍历背包
                dp[j] = min(dp[j - coins[i]] + 1, dp[j]);
            }
        }
        return dp[amount] == INF ? -1 : dp[amount];
    }
};